sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala

```diff
@@ -148,53 +150,88 @@ private[hive] class IsolatedClientLoader(
-    }
+  private[hive] var classLoader: ClassLoader = if (isolationOn) {
+    new URLClassLoader(allJars, rootClassLoader) {
+      val cache = new java.util.concurrent.ConcurrentHashMap[String, Class[_]]
+
+      override def loadClass(name: String, resolve: Boolean): Class[_] = {
+        var clazz = findLoadedClass(name)
+        if (clazz == null) {
+          clazz = cache.get(name)
+          if (clazz == null) {
+            // always assume resolve is false at this point - we'll resolve later
+            clazz = doLoadClass(name, resolve = false)
+            val clazz2 = cache.putIfAbsent(name, clazz)
+            // check if someone else beat us to updating the cache entry
+            if (clazz2 != null) {
+              // if so then we should use the cached entry to be consistent
+              clazz = clazz2
+            }
+          }
+          if (resolve) {
+            resolveClass(clazz)
+          }
+        }
+        clazz
+      }
 
-    def doLoadClass(name: String, resolve: Boolean): Class[_] = {
-      val classFileName = name.replaceAll("\\.", "/") + ".class"
-      if (isBarrierClass(name) && isolationOn) {
-        // For barrier classes, we construct a new copy of the class.
-        val bytes = IOUtils.toByteArray(baseClassLoader.getResourceAsStream(classFileName))
-        logDebug(s"custom defining: $name - ${util.Arrays.hashCode(bytes)}")
-        defineClass(name, bytes, 0, bytes.length)
-      } else if (!isSharedClass(name)) {
-        logDebug(s"hive class: $name - ${getResource(classToPath(name))}")
-        super.loadClass(name, resolve)
-      } else {
-        // For shared classes, we delegate to baseClassLoader.
-        logDebug(s"shared class: $name")
-        baseClassLoader.loadClass(name)
+      def doLoadClass(name: String, resolve: Boolean): Class[_] = {
+        val classFileName = name.replaceAll("\\.", "/") + ".class"
+        if (isBarrierClass(name)) {
+          // For barrier classes, we construct a new copy of the class.
+          val bytes = IOUtils.toByteArray(baseClassLoader.getResourceAsStream(classFileName))
+          logDebug(s"custom defining: $name - ${util.Arrays.hashCode(bytes)}")
+          defineClass(name, bytes, 0, bytes.length)
+        } else if (!isSharedClass(name)) {
+          logDebug(s"hive class: $name - ${getResource(classToPath(name))}")
+          super.loadClass(name, resolve)
+        } else {
+          // For shared classes, we delegate to baseClassLoader.
+          logDebug(s"shared class: $name")
+          baseClassLoader.loadClass(name)
+        }
       }
     }
+  } else {
+    baseClassLoader
   }
 
-  // Pre-reflective instantiation setup.
-  logDebug("Initializing the logger to avoid disaster...")
-  Thread.currentThread.setContextClassLoader(classLoader)
+  private[hive] def addJar(path: String): Unit = synchronized {
+    val jarURL = new java.io.File(path).toURI.toURL
+    classLoader = new java.net.URLClassLoader(Array(jarURL), classLoader)
```

Comment:
While you are changing this part, can you add a TODO at here saying something like we should avoid of stacking classloaders (it will be good to use a single URLClassLoader and add jars to that).